applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_fixdof_dirichlet_condition.cpp

```diff
@@ -0,0 +1,256 @@
+        }
+
+        noalias( rLeftHandSideMatrix ) = ZeroMatrix(matrix_size,matrix_size); //resetting LHS
+    }
+
+    // Resizing as needed the RHS
+    if ( CalculateResidualVectorFlag == true ) //calculation of the matrix is required
+    {
+        if ( rRightHandSideVector.size( ) != matrix_size )
+        {
+            rRightHandSideVector.resize( matrix_size, false );
+        }
+
+        noalias( rRightHandSideVector ) = ZeroVector( matrix_size ); //resetting RHS
+    }
+
+
+    // Check contact: Check contact penetration: if <0 apply constraint, otherwise no
+    this->SetValue(FIX_DOF,true);
+    if (Is(CONTACT))
+    {
+        // NOTE: the unit_normal_vector is assumed always pointing outside the boundary
+        array_1d<double, 3 > & unit_normal_vector = this->GetValue(MPC_NORMAL);
+        ParticleMechanicsMathUtilities<double>::Normalize(unit_normal_vector);
+        array_1d<double, 3 > field_displacement = ZeroVector(3);
+        for ( unsigned int i = 0; i < number_of_nodes; i++ )
+        {
+            if (Variables.N[i] > std::numeric_limits<double>::epsilon() )
+            {
+                for ( unsigned int j = 0; j < dimension; j++)
+                {
+                    field_displacement[j] += Variables.N[i] * Variables.CurrentDisp(i,j);
+                }
+            }
+        }
+
+        const double penetration = MathUtils<double>::Dot((field_displacement - imposed_displacement), unit_normal_vector);
+
+        // If penetrates, apply constraint, otherwise no
+        if (penetration >= 0.0001)
+        {
+            this->SetValue(FIX_DOF,false);
+        }
+
+    }
+
+
+    KRATOS_CATCH( "" )
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void MPMParticleFixDofDirichletCondition::FinalizeSolutionStep( ProcessInfo& rCurrentProcessInfo )
+{
+    KRATOS_TRY
+
+    MPMParticleBaseDirichletCondition::FinalizeSolutionStep(rCurrentProcessInfo);
+
+    GeometryType& r_geometry = GetGeometry();
+    const unsigned int number_of_nodes = r_geometry.PointsNumber();
+    const unsigned int dimension = GetGeometry().WorkingSpaceDimension();
+
+    for (unsigned int i = 0; i < number_of_nodes; ++i)
+        {
+            r_geometry[i].SetLock();
+            r_geometry[i].pGetDof(DISPLACEMENT_X)->FreeDof();
+            r_geometry[i].pGetDof(DISPLACEMENT_Y)->FreeDof();
+            if(dimension == 3)
+                r_geometry[i].pGetDof(DISPLACEMENT_Z)->FreeDof();
+            r_geometry[i].UnSetLock();
+        }
+
+    // Additional treatment for slip conditions
+    if (Is(SLIP))
+    {
+        // Here MPC normal vector and IS_STRUCTURE are reset
+        for ( unsigned int i = 0; i < number_of_nodes; i++ )
+        {
+            r_geometry[i].SetLock();
+            r_geometry[i].Reset(SLIP);
+            r_geometry[i].FastGetSolutionStepValue(IS_STRUCTURE) = 0.0;
```

Comment:
`IS_STRUCTUR` is deprecated
please use the `Flag` `STRUCTURE`

Reply:
I'll do that in a separate PR as there are changes for the hole application.